source/common/grpc/transcoding_filter.cc

```diff
@@ -0,0 +1,396 @@
+                 std::unique_ptr<ResponseToJsonTranslator> response_translator)
+      : request_translator_(std::move(request_translator)),
+        response_translator_(std::move(response_translator)),
+        request_stream_(request_translator_->Output().CreateInputStream()),
+        response_stream_(response_translator_->CreateInputStream()) {}
+
+  // Transcoder implementation
+  ::google::grpc::transcoding::TranscoderInputStream* RequestOutput() {
+    return request_stream_.get();
+  }
+  Status RequestStatus() { return request_translator_->Output().Status(); }
+
+  ZeroCopyInputStream* ResponseOutput() { return response_stream_.get(); }
+  Status ResponseStatus() { return response_translator_->Status(); }
+
+private:
+  std::unique_ptr<JsonRequestTranslator> request_translator_;
+  std::unique_ptr<ResponseToJsonTranslator> response_translator_;
+  std::unique_ptr<::google::grpc::transcoding::TranscoderInputStream> request_stream_;
+  std::unique_ptr<::google::grpc::transcoding::TranscoderInputStream> response_stream_;
+};
+
+} // namespace
+
+TranscodingConfig::TranscodingConfig(const Json::Object& config) {
+  std::string proto_descriptor_file = config.getString("proto_descriptor");
+  FileDescriptorSet descriptor_set;
+  if (!descriptor_set.ParseFromString(Filesystem::fileReadToEnd(proto_descriptor_file))) {
+    throw EnvoyException("Unable to parse proto descriptor");
+  }
+
+  for (const auto& file : descriptor_set.file()) {
+    if (descriptor_pool_.BuildFile(file) == nullptr) {
+      throw EnvoyException("Unable to parse proto descriptor");
+    }
+  }
+
+  google::grpc::transcoding::PathMatcherBuilder<const google::protobuf::MethodDescriptor*> pmb;
+
+  for (const auto& service_name : config.getStringArray("services")) {
+    auto service = descriptor_pool_.FindServiceByName(service_name);
+    if (service == nullptr) {
+      throw EnvoyException("Could not find '" + service_name + "' in the proto descriptor");
+    }
+    for (int i = 0; i < service->method_count(); ++i) {
+      auto method = service->method(i);
+
+      auto http_rule = method->options().GetExtension(google::api::http);
+
+      log().debug("/" + service->full_name() + "/" + method->name());
+      log().debug(http_rule.DebugString());
+
+      switch (http_rule.pattern_case()) {
+      case ::google::api::HttpRule::kGet:
+        pmb.Register("GET", http_rule.get(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPut:
+        pmb.Register("PUT", http_rule.put(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPost:
+        pmb.Register("POST", http_rule.post(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kDelete:
+        pmb.Register("DELETE", http_rule.delete_(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPatch:
+        pmb.Register("PATCH", http_rule.patch(), http_rule.body(), method);
```

Comment:
Do you have full coverage of all these cases (and the file in general)? I don't see any "PATCH" reference by a simple search, but maybe it's done some other way.